src/main/java/org/terasology/launcher/game/TerasologyGameVersions.java

```diff
@@ -83,6 +92,88 @@ public synchronized TerasologyGameVersion getGameVersionForBuildVersion(GameJob
+
+        // Get list of installed games
+        final List<TerasologyGameVersion> allInstalledGames = getInstalledGames(gameDirectory);
+
+        for (GameJob job : GameJob.values()) {
+            // Get installed games for current job
+            final List<TerasologyGameVersion> installedGames = allInstalledGames.stream()
+                    .filter(gameVersion -> gameVersion.getJob().equals(job))
+                    .collect(Collectors.toList());
+
+            final List<Integer> installedBuildNumbers = installedGames.stream()
+                    .map(TerasologyGameVersion::getBuildNumber)
+                    .collect(Collectors.toList());
+
+            // Get available games that are not installed
+            final List<TerasologyGameVersion> availableGames = packageManager.getPackageVersions(GamePackageType.byJobName(job.name()))
+                    .stream()
+                    .filter(buildNumber -> !installedBuildNumbers.contains(buildNumber))
+                    .map(buildNumber -> getGameVersion(buildNumber, job, cacheDirectory))
+                    .collect(Collectors.toList());
+
+            // Add the installed games and sort by build numbers (in descending order)
+            availableGames.addAll(installedGames);
+            availableGames.sort(Comparator.comparing(TerasologyGameVersion::getBuildNumber).reversed());
+
+            // Add extra item denoting the latest version
+            availableGames.add(0, makeLatestFrom(availableGames.get(0)));
+
+            gameVersionLists.put(job, availableGames);
+        }
+    }
+
+    private TerasologyGameVersion getGameVersion(int buildNumber, GameJob job, Path cacheDirectory) {
+        // Return cached version if it exists
+        final Path cacheFile = createCacheFile(job, buildNumber, cacheDirectory);
+        try {
+            if (Files.exists(cacheFile) && Files.isReadable(cacheFile) && Files.isRegularFile(cacheFile)) {
+                logger.debug("Found cached version for build {} of job {}", buildNumber, job);
+                try (ObjectInputStream ois = new ObjectInputStream(Files.newInputStream(cacheFile))) {
+                    return (TerasologyGameVersion) ois.readObject();
+                }
+            }
+        } catch (IOException | ClassNotFoundException e) {
+            logger.debug("Could not load cached version: {}", cacheFile.getFileName());
+        }
+
+        // Else create a new one
+        final TerasologyGameVersion gameVersion = new TerasologyGameVersion();
+        gameVersion.setJob(job);
+        gameVersion.setBuildNumber(buildNumber);
+        loadAndSetGameVersionInfo(gameVersion, null, job, buildNumber);
+        loadAndSetChangeLog(gameVersion, null, job, buildNumber);
+        loadAndSetSuccessful(gameVersion, null, job, buildNumber);
+
+        return gameVersion;
+    }
+
+    private List<TerasologyGameVersion> getInstalledGames(Path gameDirectory) {
+        try {
+            final int maxDepth = 5;
```

Comment:
Why is the search depth set to 5? Is that arbitrary, or is there a reason for that? If possible, I would like to limit it to 1…

Reply:
It's actually due to the way `GameDownloader` works and extracts the game packages. It extracts the zip in a way that the engine jar is located inside: 
`game-directory/job-installation-directory/job-name/build-number/libs`
For example, one in my laptop was:
`~/.terasology/RELEASE/TerasologyStable/78/libs`
So the engine jar can only be found at a depth of 5 from the `gameDirectory`, for now.

Reply:
Please add this as a comment to this line.
```suggestion
            // The `GameDownloader` extracts the game ZIP such that the installed game is located at depth 5
            //     `<game-directory>/<job-installation-directory>/<job-name>/<build-number>/libs
            final int maxDepth = 5;
```